pkg/provider/consulcatalog/consul_catalog.go

```diff
@@ -294,6 +352,192 @@ func contains(values []string, val string) bool {
 	return false
 }
 
+func (p *Provider) registerConnectService(ctx context.Context) {
+	if !p.ConnectAware {
+		return
+	}
+
+	ctxLog := log.With(ctx, log.Str(log.ProviderName, "consulcatalog"))
+	logger := log.FromContext(ctxLog)
+
+	if p.ServiceName == "" {
+		p.ServiceName = "traefik"
+	}
+
+	client, err := createClient(p.Endpoint)
+	if err != nil {
+		logger.WithError(err).Error("failed to create consul client")
+		return
+	}
+
+	serviceID := uuid.New().String()
+	operation := func() error {
+		regReq := &api.AgentServiceRegistration{
+			ID:   serviceID,
+			Kind: api.ServiceKindTypical,
+			Name: p.ServiceName,
+			Port: p.ServicePort,
+			Connect: &api.AgentServiceConnect{
+				Native: true,
+			},
+		}
+
+		err = client.Agent().ServiceRegister(regReq)
+		if err != nil {
+			return fmt.Errorf("failed to register service in consul catalog. %w", err)
+		}
+
+		return nil
+	}
+
+	notify := func(err error, time time.Duration) {
+		logger.Errorf("Failed to register traefik as Connect Native service in consul catalog. %s", err)
+	}
+
+	err = backoff.RetryNotify(safe.OperationWithRecover(operation), backoff.WithContext(job.NewBackOff(backoff.NewExponentialBackOff()), context.Background()), notify)
+	if err != nil {
+		logger.WithError(err).Error("failed to register traefik in consul catalog as connect native service")
+		return
+	}
+
+	<-ctx.Done()
+	err = client.Agent().ServiceDeregister(serviceID)
+	if err != nil {
+		logger.WithError(err).Error("failed to deregister traefik from consul catalog")
+	}
+}
+
+func rootsWatchHandler(ctx context.Context, dest chan<- []string) func(watch.BlockingParamVal, interface{}) {
```

Comment:
@Gufran, since you wrote most of that, any ideas?

Reply:
@mpl I fully agree with you and I just tried. But looking at the variables that are exposed I think we can only use `connect.Service` if we where to change traefik to use `Service.Dial` -- otherwise I do not think we'd have access to the certs etc… That said, I think one approach might be to copy most of the `Service` and expose what we need (ie a "verifyFunc"). This way (like you suggested for `clientSideVerifier`) we could copy in change from upstream.

Reply:
I tried and tried, but I guess unless we get consul connect to change and expose the `tlsCfg` on the service + mabye a notification channel when it changes; I think we have to copy & modify the stuff :( How good are your connections to consul to get something changed there? :D